### DISCUSSION TITLE
Best practice for initializing a longlived client at startup that requires context manager

### DISCUSSION THREAD

**Discussion by jgould22** [fastapi/fastapi#6068] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# This is how I would have done it if I did not require a context manager 
from fastapi import FastAPI

import aioboto3
from settings import get_settings

settings = get_settings()

# Create App
app = FastAPI()

# Init the boto client on start up
# Based on https://github.com/tiangolo/fastapi/issues/4035#issuecomment-941293227
@app.on_event("startup")
async def init_client() -> None:
    """
    Initialize the boto client 
    """
    boto_connection = AIOBoto3Client(<params>)
    app.dependency_overrides[AIOBoto3Client] = lambda: boto_connection

# Class to contain client 
class AIOBoto3Client:
    """Class responsible for initializing boto client channel at startup"""

    def __init__(self, host: str, port: int):
        """Init a client"""
        # Cant init a boto client here because it is a context manager 

    def get_client(self):
        """Get a client from an aioboto3 session"""
        return self.client

async def get_boto_clinet(
    boto_connection: AIOBoto3Client = Depends(),
) -> :
    """Returns the Method Stub, for use with dependency injection"""
    return boto_connection.get_client()

@app.get("/")
async def my_endpoint_with_an_aiobotoclient(clinet:aioboto3.Client = Depends(get_boto_clinet)):
    # Do stuff with my client
```

### Description

I am looking for suggestions on how to handle initializing a long lived client on start up.

Specifically I am trying to init an [aioboto3](https://github.com/terrycain/aioboto3) session/client so I can initialize it at start up and reuse it over the lifetime of the application.

In other places I have used the approach I outline here https://github.com/tiangolo/fastapi/issues/5438#issuecomment-1325321265 however that approach is a little more tricky because aioboto3 is implemented as an async context manager.

I appear to need what starlette describes as a lifespan in its events docs, where I could init my client and then yield and leave the clean up till after.

I noticed that lifespan is not exposed in fastapi (though there are some PRs that implement it). 

Does anyone have any recommendation on how I can work around the above limitation ?

### Operating System

Linux

### Operating System Details

LSB Version:	n/a
Distributor ID:	EndeavourOS
Description:	EndeavourOS Linux
Release:	rolling
Codename:	rolling

Linux node1 6.1.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 31 Dec 2022 17:40:35 +0000 x86_64 GNU/Linux

### FastAPI Version

0.88.0

### Python Version

3.11

### Additional Context

The aioboto3 repo has an issue that describes using a global (https://github.com/terrycain/aioboto3/issues/279) but I would prefer to stick to the dependency injection if possible 

Here is how I have done it in that past when initialization of the long lived client is sync and does not require clean up 
https://github.com/tiangolo/fastapi/issues/5438#issuecomment-1325321265

**Comment by ShaneNolan** [comment]
Bump because I am also interested in the best approach for this

**Comment by jgould22** [comment, marked as answer]
With the release of [0.94.0](https://github.com/tiangolo/fastapi/releases/tag/0.94.0) the lifespan start up context manager is now available so in the case of aioboto that is where I would do it now 
 
More here in the docs https://fastapi.tiangolo.com/advanced/events/#lifespan

**Comment by russoale** [comment]
I am not really an expert in FastAPI and coming from Java/Spring boot I was wondering if the following would make sense: 
```python
# main.py 
from fastapi import FastAPI

from some_async_client.api_client import AsyncApiClient

app = FastAPI(lifespan = async_client_lifespan)


# lifespan.py 
async_api_client: AsyncApiClient

@asynccontextmanager
async def async_client_lifespan(app: FastAPI) -> Any:
    global async_api_client
    async_api_client = AsyncApiClient()

    yield

    await async_api_client.close()


# service.py 
class Service:

    def __init__(self, client: typing.Optional[AsyncApiClient] = None):
        self.client = client or async_api_client

    async def do_something(self, ...) -> Any:
        # Do stuff with client


# router.py
service =   Service()

@app.get("/")
async def my_endpoint(...):
    service.do_something(...)
```

I know this is not very pythonic but my application keeps growing and becoming more complex, which is why I followed some principles from spring boot. Python's default arguments comes in very useful, which makes unit testing the service very easy, but I am not really sure if this is the best way in FastAPI to implement, or rather using FastAPI dependency injection? 
Not only in terms of testing but also resource management. 
Any help is much appreciated 🙏
